### PR TITLE
[SolYul] External function calls

### PR DESCRIPTION
**Pull request by chriseth**
Part of #8343.

### CODE REVIEW

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -743,6 +758,196 @@ bool IRGeneratorForStatements::visit(Literal const& _literal)
...
+	if (funType.gasSet())
+		templ("gas", m_context.variablePart(_functionCall.expression(), 3));
+	else if (m_context.evmVersion().canOverchargeGasForCall())
+		// Send all gas (requires tangerine whistle EVM)
+		templ("gas", "gas()");
+	else
+	{
+		// send all gas except the amount needed to execute "SUB" and "CALL"
+		// @todo this retains too much gas for now, needs to be fine-tuned.
+		u256 gasNeededByCaller = eth::GasCosts::callGas(m_context.evmVersion()) + 10;
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:917]: Why + 10?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 2]
```diff
@@ -743,6 +758,196 @@ bool IRGeneratorForStatements::visit(Literal const& _literal)
...
+		solAssert(!funType.valueSet(), "Value set for delegatecall");
+	else if (useStaticCall)
+		solAssert(!funType.valueSet(), "Value set for staticcall");
+	else if (funType.valueSet())
+		templ("value", m_context.variablePart(_functionCall.expression(), 4));
+	else
+		templ("value", "0");
+
+	bool existenceChecked = false;
+	// Check the target contract exists (has code) for non-low-level calls.
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
	// Check that the target contract exists (has code) for non-low-level calls.
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -743,6 +758,196 @@ bool IRGeneratorForStatements::visit(Literal const& _literal)
...
+		<copyDynamicReturnData>
+		<allocate>
+		mstore(<freeMem>, add(<pos>, and(add(<retSize>, 0x1f), not(0x1f))))
+		<defineRetVars> <abiDecode>(<pos>, <retSize>)
+	)");
+	string pos = m_context.newYulVariable();
+	templ("pos", pos);
+	templ("end", m_context.newYulVariable());
+	templ("result", m_context.newYulVariable());
+	templ("freeMem", to_string(CompilerUtils::freeMemoryPointer));
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
	templ("freeMem", fetchFreeMem());
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 4]
```diff
@@ -743,6 +758,196 @@ bool IRGeneratorForStatements::visit(Literal const& _literal)
...
+					", " << to_string(retSize) <<
+				"), 0)\n";
+	}
+
+	ABIFunctions abi(m_context.evmVersion(), m_context.functionCollector());
+
+	solUnimplementedAssert(!funType.isBareCall(), "");
+	Whiskers templ(R"(
+		<existenceCheck>
+		let <pos> := mload(<freeMem>)
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		let <pos> := <freeMem>
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 5]
```diff
@@ -743,6 +758,196 @@ bool IRGeneratorForStatements::visit(Literal const& _literal)
...
+	if (!m_context.evmVersion().canOverchargeGasForCall())
+	{
+		// Touch the end of the output area so that we do not pay for memory resize during the call
+		// (which we would have to subtract from the gas left)
+		// We could also just use MLOAD; POP right before the gas calculation, but the optimizer
+		// would remove that, so we use MSTORE here.
+		if (!funType.gasSet() && retSize > 0)
+			m_code <<
+				"mstore(add(" <<
+					"mload(" << to_string(CompilerUtils::freeMemoryPointer) << ")" <<
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
					fetchFreeMem() <<
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 6]
```diff
@@ -664,6 +678,7 @@ void IRGeneratorForStatements::endVisit(Identifier const& _identifier)
 	Declaration const* declaration = _identifier.annotation().referencedDeclaration;
 	if (MagicVariableDeclaration const* magicVar = dynamic_cast<MagicVariableDeclaration const*>(declaration))
 	{
+		string value;
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Not used?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:917]: As the comment says, it's the cost of `sub; call` - at least I think...

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 7]
```diff
@@ -743,6 +758,196 @@ bool IRGeneratorForStatements::visit(Literal const& _literal)
...
+
+	solUnimplementedAssert(!funType.isBareCall(), "");
+	Whiskers templ(R"(
+		<existenceCheck>
+		let <pos> := mload(<freeMem>)
+		mstore(<pos>, <shl28>(<funId>))
+		let <end> := <encodeArgs>(add(<pos>, 4) <argumentString>)
+		let <result> := <call>(<gas>, <address>, <value>, <pos>, sub(<end>, <pos>), <pos>, <retSize>)
+		if iszero(<result>) { <forwardingRevert> }
+		<copyDynamicReturnData>
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: I'll redo this with conditionals

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:917]: Wie gesagt... :p
Yea, thought so

**Review by leonardoalt — DISMISSED** [review]
Please squash

**Pull request merged**